File: write-api-server-kotlin/src/main/kotlin/com/github/j5ik2o/adceet/api/write/domain/ThreadError.kt
package com.github.j5ik2o.adceet.api.write.domain

sealed interface ThreadError {
  val message: String
  companion object {
    data class ThreadCountOverError(override val message: String) : ThreadError
    data class ExistsMemberError(val accountId: AccountId) : ThreadError {
      override val message: String = "This account($accountId) is already a member."
    }

    data class NotMemberError(val accountId: AccountId) : ThreadError {
      override val message = "This account($accountId) is not a member."
    }
    data class MessageSizeOverError(override val message: String) : ThreadError
  }
}

File: write-api-server-scala/src/main/scala/com/github/j5ik2o/adceet/api/write/domain/ThreadError.scala
package com.github.j5ik2o.adceet.api.write.domain

sealed trait ThreadError {
  def message: String
}

object ThreadError {

  final case class ThreadCountOverError(message: String) extends ThreadError
  final case class ExistsMemberError(accountId: AccountId) extends ThreadError {
    override def message: String = s"This account($accountId) is already a member."
  }
  final case class NotMemberError(accountId: AccountId) extends ThreadError {
    override def message: String = s"This account($accountId) is not a member."
  }
  final case class MessageSizeOverError(message: String) extends ThreadError

}
